pkg/controller/application/application_controller.go

```diff
@@ -280,174 +243,189 @@ func (c *Controller) processApplication(app *shipperv1.Application) error {
 		app.Spec.RevisionHistoryLimit = &max
 	}
 
-	// Clean up excessive releases regardless of exit path.
-	defer c.cleanUpReleasesForApplication(app)
+	var shouldRetry bool
 
-	latestRelease, err := c.getLatestReleaseForApp(app)
+	if err = c.processApplication(app); err != nil {
+		shouldRetry = true
+		runtime.HandleError(fmt.Errorf("error syncing Application %q (will retry): %s", key, err))
+		c.recorder.Event(app, corev1.EventTypeWarning, "FailedApplication", err.Error())
+	}
+
+	// TODO(asurikov): change to UpdateStatus when it's available.
+	_, err = c.shipperClientset.ShipperV1().Applications(app.Namespace).Update(app)
 	if err != nil {
-		app.Status.Conditions = conditions.SetApplicationCondition(
-			app.Status.Conditions,
-			shipperv1.ApplicationConditionTypeValidHistory,
-			corev1.ConditionFalse,
-			conditions.FetchReleaseFailed,
-			fmt.Sprintf("could not fetch the latest release: %q", err),
-		)
-		return err
+		shouldRetry = true
+		runtime.HandleError(fmt.Errorf("error syncing Application %q (will retry): %s", key, err))
 	}
 
-	if latestRelease == nil {
-		err = c.createReleaseForApplication(app, 0)
-		if err != nil {
-			app.Status.Conditions = conditions.SetApplicationCondition(
-				app.Status.Conditions,
-				shipperv1.ApplicationConditionTypeReleaseSynced,
-				corev1.ConditionFalse,
-				conditions.CreateReleaseFailed,
-				fmt.Sprintf("could not create a new release: %q", err),
-			)
+	return shouldRetry
+}
 
-			return err
-		}
-		app.Annotations[shipperv1.AppHighestObservedGenerationAnnotation] = "0"
-		app.Status.Conditions = conditions.SetApplicationCondition(
-			app.Status.Conditions,
-			shipperv1.ApplicationConditionTypeReleaseSynced,
-			corev1.ConditionTrue,
-			"", "",
-		)
-		return nil
+func (c *Controller) wrapUpApplicationConditions(app *shipperv1.Application, rels []*shipperv1.Release) error {
```

Comment:
This function is slightly confusing for me: I see it called as an alternate branch to code paths which also set a bunch of conditions, and this func unconditionally sets a set of conditions without doing any kind of analysis.

I think we might be in an awkward middle ground where we neither have a universal "check all the conditions and set them" func that is the last call before every return, nor do we consistently set the conditions in-line with the work. If I understand correctly, `wrapUpApplicationConditions` means "everything was OK, so mark all the conditions to reflect that and then figure out the RollingOut condition". Is that accurate? If so, I think we can probably workshop a better name; alternatively, we can think about trying to have this thing compute all the various conditions (that might need structured errors?).

Reply:
The method was created to support successful return points on the caller, informing that all the conditions that code path might report are clear; it is only executed in the case no error was found during the caller's execution (with the exception of the case where a release has been added and removed (see below). Other than that, each failure return point sets the proper conditions accordingly.

https://github.com/bookingcom/shipper/blob/bf2f0323862ae18b3da7893aeea5dd6832498f77/pkg/controller/application/application_controller.go#L378-L390

The first action point of this, I think, is to document the method accordingly. The other is finding a better name for it, which it'd be nice to hear your suggestions.